File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    unordered_map<TreeNode*, int> cache;
    int depth(TreeNode* root){
        if (!root) return 0;
        if (cache.count(root)) return cache[root];
        return cache[root] = max(depth(root->left), depth(root->right)) + 1;
    }
    bool isBalanced(TreeNode* root) {
        if (!root) return true;
        return isBalanced(root->left) && isBalanced(root->right) && abs(depth(root->left) - depth(root->right)) <= 1;
    }
};

File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = {}
        def dp(index):
            if index >= len(nums):
                return 0
            if index in cache:
                return cache[index]
            include = nums[index] + dp(index+2)
            exclude = dp(index+1)
            cache[index] = max(include, exclude)
            return cache[index]
        return max(dp(0), dp(1))
